### PR TITLE
Product Creation with AI: MainViewModel

### PR DESCRIPTION
**Pull request by hichamboushaba**
Closes: #9789 

### Description
This is a proposition for the main ViewModel to be used for the main screen, in addition to the SubScreens and their SubViewModels.

Let's discuss it in the comments.

### Testing instructions
1. Use a store eligible for product creation with AI (new WPCom store or selfhosted with Jetpack AI assistant)
2. Open the app, then from the onboarding list click on the task "Create your first product"
3. Select the AI option.
4. Confirm the new screen is shown, with a progress bar at the top (ignore the content of the sub-screen, it'll be implemented separately).

### Images/gif

- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/component/Toolbar.kt [hunk 1]
```diff
@@ -131,7 +131,7 @@ fun Toolbar(
 @Composable
 fun Toolbar(
     modifier: Modifier = Modifier,
-    title: @Composable () -> Unit = {},
+    title: @Composable () -> Unit,
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/component/Toolbar.kt:134]: This is unrelated to this task, just to resolve an ambiguous overload.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ai/AddProductWithAISubViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package com.woocommerce.android.ui.products.ai
+
+import com.woocommerce.android.viewmodel.MultiLiveEvent.Event
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.emptyFlow
+import java.io.Closeable
+
+sealed interface AddProductWithAISubViewModel<T> : Closeable {
+    val events: Flow<Event>
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ai/AddProductWithAISubViewModel.kt:9]: An optional property that subscreens can use to emit events, those events will be propagated to the main fragment; this will be useful for navigation events started from the subscreens (for example the name suggestion using AI).

We can, if we want, handle the events in the Composables themselves, but it'll be very different from the other screens, and it'll need more effort, this is why I just opted for propagating the events to the main fragment.

We'll use also the main fragment if we want to pass back some navigation result (the same example: the name suggestion bottom sheet), the main fragment will use `handleResult`, then it'll pass the result to the main ViewModel which will then pass it to the subviewModel, this is not optimal, but it's the best I could think off, if someone has a better idea, please share it.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ai/AddProductWithAISubViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+        get() = emptyFlow()
+    val onDone: (T) -> Unit
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ai/AddProductWithAISubViewModel.kt:11]: An callback that subscreens can use to notify the parent screen of completion, and potentially pass data to it.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ai/AddProductWithAIViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    )
+
+    enum class SaveButtonState {
+        Hidden, Shown, Loading
+    }
+}
+
+private enum class Step {
+    ProductName, AboutProduct, Preview
+}
```
Comment (hafizrahman) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ai/AddProductWithAIViewModel.kt]: possible np: Will it improve readability/understandability if we're more explicit about the step number for each screen? Perhaps also avoid weirdness if someone accidentally changes the order of the enum members.

```
private enum class Step(val stepNumber: Int) {
    ProductName(1),
    AboutProduct(2),
    Preview(3)
}
```

(will also require changing `step.value.ordinal` to `step.value.stepNumber`).

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ai/AddProductWithAISubViewModel.kt (hunk 2 above)
Comment (hafizrahman) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ai/AddProductWithAISubViewModel.kt:9]: > then it'll pass the result to the main ViewModel which will then pass it to the subviewModel, this is not optimal

Would it make it easier to have `suggestedProductName` string as part of the main viewmodel's state? `handleResult` can update its value, then we just pass it as parameter of `ProductNameSubScreen()` and avoid having to use the sub-viewmodel.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ai/AddProductWithAIScreen.kt [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+                }
+            )
+        }
+    ) {
+        Column(
+            modifier = Modifier
+                .fillMaxSize()
+                .padding(it)
+        ) {
+            LinearProgressIndicator(progress = state.progress, modifier = Modifier.fillMaxWidth())
```
Comment (hafizrahman) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ai/AddProductWithAIScreen.kt:67]: nice way to persist progress state between sub-screens ✨

**Comment by hafizrahman** [comment]
Overall looks good to me 👍🏼 , just left some quick comments.

**Comment by JorgeMucientes** [comment]
Thanks for putting this together @hichamboushaba. I like the current approach and I think is very flexible. 

However, I can't stop thinking about how we might be overengineering things a bit. Considering the following: 
- Most of the "complex" logic will be handled by bottom sheets in their independent fragments and viewmodels
- `AddProductWithAIViewModel` handles: results data from bottom sheets, navigation between steps, and progress bar state. Doesn't seem like a lot. In fact, there's more code needed to handle the `subviewmodels` logic than to handle the rest of the state. 
- Judging from the designs, all steps are mostly static screens with barely any logic, it seems like all the subview-models are essentially going to be “empty dummy” classes with almost nothing in them. Only the `Preview` step will have some logic. But i'm not sure if `Preview` step has enough logic to justify going for this subViewModel approach. 

I understand the main reason for this design is to enable us to work in parallel easily, plus getting some benefit of extracting and encapsulating some logic into subviewmodels. But at the cost of making code more complex imho, and adding extra "boilerplate" to support the subviewmodel handling from the main viewmodel class. 

I think that if we can define now a general UI state model with all we need for the whole flow, we'd be okay using a single `AddProductWithAIViewModel`and working over this flow in parallel. 

I think something like this could work well for whole UI state: 

```kotlin
    data class State(
        val currentStep: Step, // I think this should be enough to determine any Compose UI specific changes
        val progress: Float,
        val aiProductName: String,
        val aiProductDescription:String,
        val aiTone: AiTone,
        val isLoadingPreview: Boolean,
        val saveButtonState: SaveButtonState
    )

    enum class SaveButtonState {
        Hidden, Shown, Loading
    }

    enum class AiTone{
        Casual,
        Formal,
        Flowery,
        Convincing
    }
```

Wdyt folks? @hichamboushaba @hafizrahman

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ai/AddProductWithAISubViewModel.kt (hunk 2 above)
Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ai/AddProductWithAISubViewModel.kt:9]: > Would it make it easier to have suggestedProductName string as part of the main viewmodel's state? handleResult can update its value, then we just pass it as parameter of ProductNameSubScreen() and avoid having to use the sub-viewmodel.

Personally I think it's still more clean to pass it through the `subviewModel`, and for two reasons:
1. Better separation of concerns.
2. It'll actually make everything work nicely, as the generated name will directly update the subviewmodel's state, and then it'll be returned in the `onDone` callback.

A different approach if we want to avoid this: Instead of making the `AIProductNameBottomSheet` as a separate dialog, just handle it in Compose, using `ModalBottomSheetLayout`, then use the `subviewmodel` to handle its visibility, and its result, without involving the `Navigation` component at all, and it'll save us some code, as instead of creating a separate ViewModel, then we can handle the logic inside it. The downside is that we won't be able to reuse the dialog with the Product name/description feature in the other area, but IMO it's still worth it, WDYT?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ai/AddProductWithAIViewModel.kt (hunk 4 above)
Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ai/AddProductWithAIViewModel.kt]: Good point, I'll make the change 🙂

Comment (JorgeMucientes) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ai/AddProductWithAIViewModel.kt]: I'm curious why you added this as a private class outside `AddProductWithAIViewModel` ?

Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ai/AddProductWithAIViewModel.kt]: I actually just moved it inside the class in the last commit [895f6af](https://github.com/woocommerce/woocommerce-android/pull/9807/commits/895f6af52f9b4182cf78f898c853032a74fce85a) 😆

Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ai/AddProductWithAIViewModel.kt]: Done in https://github.com/woocommerce/woocommerce-android/pull/9807/commits/895f6af52f9b4182cf78f898c853032a74fce85a

**Comment by hafizrahman** [comment]
Since we've discussed this together, let's merge this so we can continue with the subscreen works. 👍🏼

**Comment by hichamboushaba** [comment]
I forgot to put a "not ready for merge" label 🤦, sorry, this merge will make reviewing the code of the previous PR harder, I'll inform Jorge about it 🙂.

@hafizrahman if you want to start working on a new task, just create the branch based on the branch `issue/9790-product-creation-ai-bottomsheet` which was the base of the PR, it'll have all the changes you need to move forward.
